### PR TITLE
Api permissions

### PR DESCRIPTION
**Pull request by markusahlstrand**
Before we only validated that a user had a token. With this change a user is made an admin when they create a tenant and all endpoints are secured so that only admins of the connected tenant has access. The next step is to add some mechanic to add additional admins or having some permission that can be used to manage all tenants.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,28 +1,28 @@
 # Cloudworker auth
 
-This is work in progress and not yet ready for even testing.
+This is a work in progress and not yet ready for even testing.
```
Comment (markusahlstrand) [inline README.md:3]: Grammerly had some feedback on the docs :)

File: src/migrate/migrations/2022-12-11T09:17:35_init.ts [hunk 2]
```diff
@@ -77,7 +82,7 @@ export async function up(db: Kysely<any>): Promise<void> {
 }
 
 export async function down(db: Kysely<any>): Promise<void> {
-  await db.schema.dropTable("users_tenants").execute();
```
Comment (markusahlstrand) [inline src/migrate/migrations/2022-12-11T09:17:35_init.ts:80]: Before we had a mapping between users and tenants intended for permissions. Now we have a separate user-table only for admins as they are different things compared to the users in the tenants. In this case we use a different auth for management.

File: src/routes/tsoa/applications.ts [hunk 3]
```diff
@@ -19,6 +19,28 @@ import { getDb } from "../../services/db";
...
+import { Context } from "cloudworker-router";
+import { Env } from "../../types";
+
+async function checkAccess(ctx: Context<Env>, tenantId: string, id: string) {
+  const db = getDb(ctx.env);
+
+  const application = await db
+    .selectFrom("applications")
+    .innerJoin("tenants", "tenants.id", "applications.tenantId")
+    .innerJoin("admin_users", "tenants.id", "admin_users.tenantId")
```
Comment (markusahlstrand) [inline src/routes/tsoa/applications.ts:32]: Join on admin_users. Maybe we should check the role here as well..

File: src/routes/tsoa/applications.ts [hunk 4]
```diff
@@ -72,6 +104,9 @@ export class ApplicationsController extends Controller {
     const { env } = request.ctx;
 
     const db = getDb(env);
+
+    await checkAccess(request.ctx, tenantId, id);
```
Comment (markusahlstrand) [inline src/routes/tsoa/applications.ts:108]: A small helper to check that the user has permissions before we do the update.

File: src/routes/tsoa/tenants.ts [hunk 5]
```diff
@@ -9,10 +9,10 @@ import {
   SuccessResponse,
   Security,
 } from "@tsoa/runtime";
-import { Tenant } from "../../types/sql/Tenant";
+import { Tenant, AdminUser } from "../../types/sql";
 import { getDb } from "../../services/db";
 import { RequestWithContext } from "../../types/RequestWithContext";
-import { v4 as uuidv4 } from "uuid";
+import { nanoid } from "nanoid";
```
Comment (markusahlstrand) [inline src/routes/tsoa/tenants.ts:15]: Ah.. forgot about this. Think nanoid is prettier. Had an issue with running nanoid with cloudflare before but think we solved it somehow.

File: src/routes/tsoa/tenants.ts [hunk 6]
```diff
@@ -35,15 +42,29 @@ export class TenantsController extends Controller {
...
-      id: uuidv4(),
+      id: nanoid(),
+      createdAt: new Date().toISOString(),
+      modifiedAt: new Date().toISOString(),
+    };
+
+    const adminUser: AdminUser = {
+      id: ctx.state.user.sub,
+      // TODO: Fetch this from the profile endpoint
+      email: "placeholder",
```
Comment (markusahlstrand) [inline src/routes/tsoa/tenants.ts:58]: This is a good question.. The email is not part of the access-token. We could query a profile endpoint to get this? Or maybe use the id-token? Or we make some email verification thing?

File: src/routes/tsoa/users.ts [hunk 7]
```diff
@@ -9,27 +9,58 @@ import {
...
+    .executeTakeFirst();
+
+  if (!user) {
+    // Application not found. Could be that the user has no access
+    throw new NotFoundError();
+  }
+}
 
 @Route("tenants/{tenantId}/users")
+@Security("oauth2", [])
```
Comment (markusahlstrand) [inline src/routes/tsoa/users.ts:46]: Missed this.. So the access token wasn't parsed.

File: README.md (hunk 1 above)
Comment (danielgent) [reply README.md:3]: I've always wanted to have it on CI/CD :smile:

File: src/routes/tsoa/applications.ts [hunk 8]
```diff
@@ -29,11 +51,16 @@ export class ApplicationsController extends Controller {
...
-    const db = getDb(request.ctx.env);
+    const { ctx } = request;
+
+    const db = getDb(ctx.env);
     const applications = await db
       .selectFrom("applications")
-      .where("applications.tenantId", "=", tenantId)
-      .selectAll()
+      .innerJoin("tenants", "tenants.id", "applications.tenantId")
+      .innerJoin("admin_users", "tenants.id", "admin_users.tenantId")
```
Comment (danielgent) [inline src/routes/tsoa/applications.ts:60]: How performant are all these joins? 

I suppose we aren't going to have that many tenants, or that many admin users, and each tenant won't have that many apps? :smile: 

image

File: src/types/sql/AdminUser.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+export interface AdminUser {
+  id: string;
+  email?: string;
+  tenantId: string;
+  createdAt: string;
+  modifiedAt: string;
```
Comment (danielgent) [inline src/types/sql/AdminUser.ts:6]: :smiling_face_with_tear: The inconsistency of camel vs snake is so painful all the time :joy:

File: src/routes/tsoa/tenants.ts (hunk 6 above)
Comment (danielgent) [reply src/routes/tsoa/tenants.ts:58]: It feels like the email would be useful in the access token but then the `sub` is unique enough...

Maybe look the email up when needed? :shrug:

**Review by danielgent — APPROVED** [review]

File: src/routes/tsoa/applications.ts (hunk 8 above)
Comment (markusahlstrand) [reply src/routes/tsoa/applications.ts:60]: The joins are fast since it's all primary keys. Would work fine with millions of vendors and apps.

**Pull request merged**